### ISSUE TITLE
Experiment with QQuickGraphicsDevice

### ISSUE THREAD

**Issue by nolankramer** [vsg-dev/vsgQt#39]
In Qt6, Qt has added `QQuickGraphicsDevice::fromDeviceObjects`, which allow referencing Vulkan resources to perform drawing: https://doc.qt.io/qt-6/qquickgraphicsdevice.html#fromDeviceObjects

This seems like far better integration with the Qt rendering stack - rather than creating a native window and embedding it (which has a huge list of issues, such as not being able to stack widgets on top, resizing lag, etc etc).

Using `QQuickGraphicsDevice` should allow drawing directly into a texture that is owned by Qt, which will then be able to display it seamlessly with other widgets.

**Comment by robertosfield** [comment]
Originally vsgQt used to use Qt's Vulkan integration but I had to abandon that code path as it wouldn't scale to efficient multi-window support for Vulkan.  Qt ties the vkDevice per Window forcing applications to duplicate data per Window, something that very rapidly becomes a major memory and performance bottleneck for applications with large models and the multiple windows. 

To address this fundamental problem in the design/implementation of Qt's Vulkan integration I had to pull out all the original Qt Vulkan integration code and use the VSG's native support for creating the required Vulkan surface, instance, physical device and logical device.   This change enables Qt/VulkanSceneGraph applications to be able to scale in the same way that VulkanSceneGraph applications on other Windowing toolkits can.

The PR that introduce this change:
https://github.com/vsg-dev/vsgQt/pull/27/files

The write up explaining the issues and how I resolved them:
https://github.com/vsg-dev/VulkanSceneGraph/discussions/862

The bottom line is Trolltech's support for Vulkan is shoehorned to work like OpenGL which might make it easier for them to manage both Vulkan and OpenGL backends with one codebase but it nobbles Vulkan in the process, and VulkanSceneGraph with it.  Both Vulkan and VulkanSceneGraph focus is on supporting hardware in the most efficient way possible, Qt's focus is quite different.

Now if you wanted to implement your own Qt/VulkanSceneGraph integration you are free to do so, but expect limitations in being able to fully control Vulkan and getting the best out of the hardware and VulkanSceneGraph.  If Qt is your focus and graphics is secondary then might be fine.

**Comment by nolankramer** [comment]
Hey @robertosfield, do you mind if we re-open this?

What I'm proposing to experiment with is a different track than what you were originally doing.

Qt's [example](https://doc.qt.io/qt-6/qtgui-hellovulkanwidget-example.html), and the implementation you have here, are cut of the same cloth - they both rely on `QWidget::createWindowContainer`.

This method has all of the downsides of `createWindowContainer`, specifically those listed [here](https://doc.qt.io/qt-6/qwidget.html#createWindowContainer). One of which is an ugly resize lag, inability to stack widgets, and Qt render perf issues depending on the number of native windows created this way (I believe due to synchronizing event loops).
It's really quite an ugly hack, but it's easy to do, so it's used in a lot of use-cases.

---

What I'm proposing is more of a double-buffer track - where VSG renders to an off-screen texture which is rendered into a QQuickWidget. The setup would look like this:
1. Set the QQuickWindow graphics API to Vulkan with `QQuickWindow::setGraphicsApi`
2. Create a Qt graphics device **from our vkDevice** with `QQuickGraphicsDevice::fromDeviceObjects`
3. Set the QQuickWindow graphics device with the above created graphics device
4. Subclass `QQuickWidget` and have it create a `QSGTexture` that in-turn delivers a `QRhiTexture`. The underlying texture buffer can be accessed with `QRhiTexture::nativeTexture()`
5. VSG renders into the above buffer
6. The `QQuickWidget` renders a scene with a single `QSGSimpleTextureNode`, which ultimately renders the `QSGTexture` that is "shared" with VSG
7. There are various events that can be used to integrate the render loop, such as `QQuickItem::updatePaintNode` or `QQuickWindow::beforeRendering`

All of this, to avoid this dreadful practice of embedding and all of it's downsides.

The upsides of this are many:
- Native widget rendering, which speeds up Qt and application responsiveness
- Integration with Qt Quick, which opens the doors to QML-based VSG widgets
- QQuickWidgets can be stacked, so other Qt Widgets can be drawn on top of them
- Multiple windows **and multiple widgets** can employ this method, and use the same vkDevice owned by VSG
- vsgQt API can be extended to support this method, with subclasses and utility methods provided to customize and make it easy for users

Let me know what your thoughts here are. I do understand what I've proposed is complicated, and your understanding of Vulkan and VSG is far greater than mine.

With this being a very complicated endeavor, I'm happy to do the experiment myself - just re-open the issue and I'll get started.

**Comment by robertosfield** [comment]
I closed this ticket item because it's not a bug with the current vsgQt, but a desire for you to use a different approach for Qt/VSG integration. This type of open ended discussion is perfect fine just not clogging up an issue tracker, issue trackers really just get seen by a small number of folks, the place for general VSG discussion is:

https://github.com/vsg-dev/VulkanSceneGraph/discussions

What you are proposing is a Qt6 only solution so is only appropriate for the subset of Qt users that have adopted Qt6, while vsgQt works on both Qt5 and Qt6, and even works on Qt5 version prior to the Qt support for Vulkan.  I don't see any harm in having a vsgQt6 library or vsgQtQuick if that's your preference.  I don't personally have a lot of Qt expertise - I have learned what I need to to learn, but my day job is very much VSG/Vulkan so if you want to pick up the baton and come up with your own more Qt6 focused library then you are welcome to,

**Comment by jacky-wang7** [comment]
> 嘿，您介意我们重新开放吗？
> 
> 我建议尝试的是一个与你最初做的不同的轨道。
> 
> Qt [的示例](https://doc.qt.io/qt-6/qtgui-hellovulkanwidget-example.html)和您在这里的实现是相同的 - 它们都依赖于 .`QWidget::createWindowContainer`
> 
> 这种方法具有 的所有缺点，特别是[此处](https://doc.qt.io/qt-6/qwidget.html#createWindowContainer)列出的缺点。其中之一是丑陋的调整大小滞后、无法堆叠小部件以及 Qt 渲染性能问题，具体取决于以这种方式创建的本机窗口的数量（我相信是由于同步事件循环）。这确实是一个相当丑陋的 hack，但它很容易做到，因此它被用于很多用例。`createWindowContainer`
> 
> 我提议的更像是一个双缓冲区轨道 - 其中 VSG 渲染为屏幕外纹理，该纹理被渲染为 QQuickWidget。设置将如下所示：
> 
> 1. 将 QQuickWindow 图形 API 设置为 Vulkan，并带有`QQuickWindow::setGraphicsApi`
> 2. 使用我们的 **vkDevice** 创建一个 Qt 图形设备`QQuickGraphicsDevice::fromDeviceObjects`
> 3. 使用上面创建的图形设备设置 QQuickWindow 图形设备
> 4. 子类，并让它创建一个，而该子类又会传递一个 .底层纹理缓冲区可以通过`QQuickWidget``QSGTexture``QRhiTexture``QRhiTexture::nativeTexture()`
> 5. VSG 渲染到上述缓冲区
> 6. 使用单个 渲染场景，最终渲染与 VSG “共享” 的场景`QQuickWidget``QSGSimpleTextureNode``QSGTexture`
> 7. 有多种事件可用于集成渲染循环，例如 或`QQuickItem::updatePaintNode``QQuickWindow::beforeRendering`
> 
> 所有这一切，都是为了避免这种可怕的嵌入做法和它的所有缺点。
> 
> 这样做的好处很多：
> 
> * 原生小部件渲染，可加快 Qt 和应用程序的响应速度
> * 与 Qt Quick 集成，为基于 QML 的 VSG 小部件打开了大门
> * QQuickWidgets 可以堆叠，因此可以在它们上面绘制其他 Qt Widgets
> * 多个窗口**和多个小部件**可以采用此方法，并使用 VSG 拥有的相同 vkDevice
> * vsgQt API 可以扩展以支持此方法，并提供子类和实用程序方法来自定义并方便用户使用
> 
> 让我知道你的想法。我确实理解我提议的内容很复杂，而您对 Vulkan 和 VSG 的理解远比我深刻。
> 
> 由于这是一项非常复杂的工作，我很乐意自己做这个实验 - 只需重新打开问题，我就可以开始了。

hi,
I think this is a good solution. Stacking widgets is very common in QT development. The problem of not being able to stack widgets has troubled me for a long time. Thank you very much for the ideas you provided. I will try to develop it myself.
